Sources/StevensonCore/SlackCommandHandler.swift

```diff
@@ -0,0 +1,87 @@
+import Vapor
+
+public typealias SlackCommand = (
+    help: String,
+    token: String,
+    parser: (SlackCommandMetadata) throws -> Command
+)
+
+public struct SlackCommandMetadata: Decodable {
+    public let token: String
+    public let channelName: String
+    public let text: String
+}
+
+public struct SlackCommandHandler: CommandHandler {
+    public enum Error: Swift.Error, Debuggable {
+        case unknownCommand(String)
+        case invalidCommand(String)
+
+        public var identifier: String {
+            return ""
+        }
+
+        public var reason: String {
+            switch self {
+            case let .unknownCommand(cmd):
+                return "Unknown command `\(cmd)`."
+            case let .invalidCommand(reason):
+                return "Invalid command: \(reason)"
+            }
+        }
+    }
+
+    let registry: [String: SlackCommand]
+    let requireChannel: String?
+
+    public init(commands: [String: SlackCommand], requireChannel: String?) {
+        self.registry = commands
+        self.requireChannel = requireChannel
+    }
+
+    public func command(from request: HTTPRequest) throws -> Command {
+        let name = request.url.path
+        let decoder = JSONDecoder()
+        decoder.keyDecodingStrategy = .convertFromSnakeCase
+        let data = request.body.data ?? Data()
+        let content = try decoder.decode(SlackCommandMetadata.self, from: data)
+        guard let reg = registry[name] else {
+            throw Error.unknownCommand(name)
+        }
+        if content.token != reg.token {
+            throw Error.invalidCommand("Invalid token")
+        }
+        if let requireChannel = requireChannel, content.channelName != requireChannel {
+            throw Error.invalidCommand("Invalid channel")
+        }
+
+        if content.text == Command.helpCommandName {
+            return Command.help(text: reg.help)
+        }
+        return try reg.parser(content)
+    }
+
+    public func result(from response: String) throws -> HTTPResponse {
+        return HTTPResponse(
+            status: .ok,
+            body: try JSONEncoder().encode([
+                "response_type": "in_channel",
+                "text": response
+                ])
+        )
+    }
+
+}
+
+public extension Command {
+    public static let helpCommandName = "help"
+    public static let helpCommandText = "helpText"
+
+    public static func help(text: String) -> Command {
+        return Command(
+            name: Command.helpCommandName,
+            arguments: [Command.helpCommandText: text],
+            branch: nil
+        )
```

Comment:
It's not clear why would we create a CI `Command` for `help` here.
Reading just this it seems like this will then trigger a CircleCI command here, as `Command` is designed as a struct used for triggering CI commands (see other comment). Even though after careful look I see on `routes.swift:16` that this command goes into a particular treatment as an exception case, this does not seem very readable to me. Had to scroll in various parts of the code to realize that it was treated kind of as a workaround in routes.

---

I'd suggest instead to have some more understandable abstraction around all this concept of `Command` here, where we'd be able to create some abstract `Command` (or call it `protocol Action`?).

For example we could then either return from `action(from request: HTTPRequest) throws -> Action` a `CICommand: Action` (for the common cases where we actually want to run some action on CI) or a `Action.noop(response:)` (for this help case where we don't want any side-effect but directly a response)

Or another alternative would be to group together `command(from request: HTTPRequest)`  and `result(from:)` to make it return a `Future<HTTPResponse>` directly, where `help` would return a `.just` while other commands triggering some CI action would `flatMap` to the proper CICommand first (would requires to inject the `ci` but we _kinda_ depend on the concept of a CI command anyway given how `struct Command` is structured...